readme.md
# esphome-q7rf

This is an ESPHome custom component which allows you to control a Computherm/Delta Q7RF/Q8RF receiver equiped furnace using a TI CC1101 transceiver module. It defines a switch platform for state toggling and a service for pairing.

I've tested this project with an ESP8266 module (NodeMCU). It should work with the ESP32 as well, since protocol timing critical part is done by the CC1101 modem.

Current tested compatible ESPHome version: v2021.10.3

**Use this project at your own risk. Reporting and/or fixing issues is always welcome.**

## Hardware
You need a CC1101 module which is assembled for 868 MHz. The chip on its own can be configured for many targets, but the antenna design on the board have to be tuned for the specific frequency in mind.

The CC1101 module is connected to the standard SPI pins of ESP8266 (secondary SPI PINs, the first set is used by the SPI flash chip).

Connections:

[CC1101 868MHz module pinouts](./doc/cc1101-pinout.jpg)

[NodeMCU module pinouts](./doc/nodemcu.jpg)

    NODEMCU               CC1101
    ============================
    3.3V                  VCC
    GND                   GND
    D7 (GPIO13/HMOSI)     MOSI
    D5 (GPIO14/HSCLK)     SCLK
    D6 (GPIO12/HMISO)     MISO
    D8 (GPIO15/HCS)       CSN

## ESPHome setup

If you're not familiar with ESPHome and its integration with Home Assistant, please read the relevant section in the [official manual](https://esphome.io/guides/getting_started_hassio.html).

Add this component using the following configuration in your node's yaml file:

    external_components:
      - source: github://nistvan86/esphome-q7rf@main
        components: [ q7rf ]

    switch:
      - platform: q7rf
        name: Q7RF switch
        cs_pin: D8
        q7rf_device_id: 0x6ed5
        q7rf_resend_interval: 60000
        q7rf_turn_on_watchdog_interval: 0

    spi:
      clk_pin: D5
      miso_pin: D6
      mosi_pin: D7

Where:
* `q7rf_device_id` (required): is a 16 bit transmitter specific ID and learnt by the receiver in the pairing process. If you operate multiple furnaces in the vicinity you must specify unique IDs for each transmitter. You can generate random identifiers for example with [random-hex](https://www.browserling.com/tools/random-hex) (use 4 digits).

* `q7rf_resend_interval` (optional): specifies how often to repeat the last state in milliseconds. Since this is a simplex protocol, there's no response arriving from the receiver and we need to compensate for corrupt or lost messages by repeating them.

  Default is: 60000 ms (1 minute)

* `q7rf_turn_on_watchdog_interval` (optional): specifies how long the furnace can stay turned on after the last `write_state` call arrived for the switch component in milliseconds. This can be used for example in conjunction with the `keep_alive` setting of Home Assistant's [generic thermostat](https://www.home-assistant.io/integrations/generic_thermostat/) component to add an additional safe guard against the crash of HA and to prevent excessive heating costs.

  Default is: 0 ms (no watchdog).

  Example: 900000 ms (15 minutes) and generic thermostat `keep_alive` set to 3 minutes

Once flashed, check the logs (or the UART output of the ESP8266) to see if configuration was successful.

You should see similar lines (note: C/D lines are only visible if you left the logger's level at the default DEBUG or lower):

During the `setup()` initialization:

    [I][q7rf.switch:x]: CC1101 initialized.

During configuration print:

    [C][q7rf.switch:x]: Q7RF:
    [C][q7rf.switch:x]:   CC1101 CS Pin: GPIO15
    [C][q7rf.switch:x]:   Q7RF Device ID: 0x6ed5
    [C][q7rf.switch:x]:   Q7RF Resend interval: 60000 ms
    [C][q7rf.switch:x]:   Q7RF Turn on watchdog interval: 0 ms

In Home Assistant under _Configuration_ → _Entities_ you should see a new switch with the same name you have specified ("Q7RF switch" in this example). In case you have disabled the automatic dashboard, add the switch to one of your dashboards. Find it and try toggling it. In the ESPHome log output you should see the component reacting:

    [D][switch:x]: 'Q7RF switch' Turning ON.
    [D][switch:x]: 'Q7RF switch': Sending state ON
    [D][q7rf.switch:x]: Handling prioritized message.
    [D][q7rf.switch:x]: Sending message: HEAT ON

## Pairing with the receiver

In order to make the receiver recognize the transmitter, we need to execute the pairing process.

Go to Home Assistant's _Developer tools_ → _Services_ and select the service `esphome.<NODE_NAME>_q7rf_pair`. Press and hold the M/A button on the receiver until it starts flashing green. Now press _Call service_ in the _Services_ page. ESPHome log will show a similar output:

    [I][q7rf.switch:x]: Enqueued pairing.
    [D][q7rf.switch:x]: Handling prioritized message.
    [D][q7rf.switch:x]: Sent message: PAIR

The receiver should stop flashing, and the pairing is now complete. Test if the associated Home Assistant switch controls the furnace correctly by manually toggling it.

## Usage example

You can configure Home Assistant's [generic thermostat](https://www.home-assistant.io/integrations/generic_thermostat/) to control the furnace (use the new switch as the `heater`).

## Q&A

**How can I revert back to the bundled wireless thermostat?**

Set the receiver into learning mode with the M/A button, then hold the SET + DAY button on your wireless thermostat until the blinking stops.

**Can I use this simultaneously with the wireless thermostat the receiver came with?**

No. Both transmitters should have their own device IDs. The receiver only listens to the currently paired device. Even if you manage to clone the wireless thermostat's ID, transmitters will not know about each other and switch the furnace in a hectic way (especially with the message repeating in place).

Since the CC1101 is a transreceiver it would be possible to extend this project to allow pairing the wireless thermostat to this ESPHome component (using a separate device ID pair), then listen and forward it's messages. In my opinion the protocol is too limited to do anything useful with the original thermostat in a home automation environment: no temperature or scheduling is transmitted, only on/off signal is sent. It's better to replace it completely.

## Resources

* The [cc1101-ook library](https://github.com/martyrs/cc1101-ook) which functioned as a template for the communication best practices with the modem.
* [denx's awesome article series](https://ardu.blog.hu/2019/04/17/computherm_q8rf_uj_kihivas_part) about reverse engineering the Q8RF's protocol. Unfortunatelly it's only
  available in Hungarian.
* [CC1101 product manual](http://www.ti.com/lit/ds/symlink/cc1101.pdf) from TI


changelog.md
# Changelog

## 2021-11-09
- Added watchdog functionality
- Reworked into external component structure
- Bugfixes reported by 2225damjaich42 and Tamás Ficzere
    - don't strict check modem version (new variant of CC1101 appeared in the wild)
    - better pair service name
- Log sent command name instead of hex value
- Updated readme
- Added changelog

## 2021-11-04
- Updated for ESPHome v2021.10.3

## 2021-01-01
- Initial commit.


components/q7rf/q7rf.h
#pragma once

#include "esphome/core/component.h"
#include "esphome/components/switch/switch.h"
#include "esphome/components/spi/spi.h"
#include "esphome/components/api/custom_api_device.h"

namespace esphome {
namespace q7rf {

static const uint8_t MSG_NONE = 0;
static const uint8_t MSG_HEAT_ON = 1;
static const uint8_t MSG_HEAT_OFF = 2;
static const uint8_t MSG_PAIR = 3;

class Q7RFSwitch : public switch_::Switch,
                   public api::CustomAPIDevice,
                   public PollingComponent,
                   public spi::SPIDevice<spi::BIT_ORDER_MSB_FIRST, spi::CLOCK_POLARITY_LOW, spi::CLOCK_PHASE_LEADING,
                                         spi::DATA_RATE_1KHZ> {
 protected:
  uint16_t q7rf_device_id_ = 0;
  uint32_t q7rf_resend_interval_ = 60000;
  uint32_t q7rf_turn_on_watchdog_interval_ = 0;

  bool initialized_ = false;
  uint8_t msg_pair_[45];
  uint8_t msg_heat_on_[45];
  uint8_t msg_heat_off_[45];
  bool state_ = false;
  uint8_t pending_msg_ = MSG_NONE;
  unsigned long last_msg_time_ = 0;
  unsigned long last_turn_on_time_ = 0;
  uint8_t msg_errors_ = 0;

 private:
  bool reset_cc();
  void send_cc_cmd(uint8_t cmd);
  void read_cc_register(uint8_t reg, uint8_t *value);
  void read_cc_config_register(uint8_t reg, uint8_t *value);
  void write_cc_register(uint8_t reg, uint8_t *value, size_t length);
  void write_cc_config_register(uint8_t reg, uint8_t value);
  bool send_cc_data(const uint8_t *data, size_t length);

  bool send_msg(uint8_t msg);
  void set_state(bool state);

 public:
  Q7RFSwitch() : PollingComponent(1000) {}
  void setup() override;
  void write_state(bool state) override;
  void dump_config() override;
  void update() override;

  void set_q7rf_device_id(uint16_t id);
  void set_q7rf_resend_interval(uint32_t interval);
  void set_q7rf_turn_on_watchdog_interval(uint32_t interval);
  void on_pairing();
};

}  // namespace q7rf
}  // namespace esphome


components/q7rf/switch.py
import esphome.codegen as cg
import esphome.config_validation as cv
from esphome.components import switch
from esphome.components import spi
from esphome.const import CONF_ID

DEPENDENCIES = ["spi"]

CONF_Q7RF_DEVICE_ID = "q7rf_device_id"
CONF_Q7RF_RESEND_INTERVAL = "q7rf_resend_interval"
CONF_Q7RF_TURN_ON_WATCHDOG_INTERVAL = "q7rf_turn_on_watchdog_interval"

q7rf_ns = cg.esphome_ns.namespace("q7rf")
Q7RF_SWITCH = q7rf_ns.class_("Q7RFSwitch", switch.Switch, cg.Component, spi.SPIDevice)

CONFIG_SCHEMA = (
    switch.SWITCH_SCHEMA.extend({cv.GenerateID(): cv.declare_id(Q7RF_SWITCH)})
    .extend(cv.COMPONENT_SCHEMA)
    .extend(spi.spi_device_schema(cs_pin_required=True))
    .extend(
        cv.Schema(
            {
                cv.Required(CONF_Q7RF_DEVICE_ID): cv.hex_uint16_t,
                cv.Optional(CONF_Q7RF_RESEND_INTERVAL): cv.uint32_t,
                cv.Optional(CONF_Q7RF_TURN_ON_WATCHDOG_INTERVAL): cv.uint32_t,
            }
        )
    )
)


def to_code(config):
    var = cg.new_Pvariable(config[CONF_ID])
    yield cg.register_component(var, config)
    yield switch.register_switch(var, config)
    yield spi.register_spi_device(var, config)

    cg.add(var.set_q7rf_device_id(config[CONF_Q7RF_DEVICE_ID]))
    if CONF_Q7RF_RESEND_INTERVAL in config:
        cg.add(var.set_q7rf_resend_interval(config[CONF_Q7RF_RESEND_INTERVAL]))
    if CONF_Q7RF_TURN_ON_WATCHDOG_INTERVAL in config:
        cg.add(
            var.set_q7rf_turn_on_watchdog_interval(
                config[CONF_Q7RF_TURN_ON_WATCHDOG_INTERVAL]
            )
        )


components/q7rf/q7rf.cpp
#include "esphome/core/log.h"
#include "q7rf.h"

namespace esphome {
namespace q7rf {

static const char *TAG = "q7rf.switch";

static const char *Q7RF_PREAMBLE_DATA = "111000111";
static const char *Q7RF_ZERO_BIT_DATA = "011";
static const char *Q7RF_ONE_BIT_DATA = "001";
static const char *Q7RF_GAP_DATA = "000";

static const uint8_t Q7RF_MSG_CMD_PAIR = 0x00;
static const uint8_t Q7RF_MSG_CMD_TURN_ON_HEATING = 0xFF;
static const uint8_t Q7RF_MSG_CMD_TURN_OFF_HEATING = 0x0F;

static const uint8_t CMD_SRES = 0x30;
static const uint8_t CMD_STX = 0x35;
static const uint8_t CMD_SIDLE = 0x36;
static const uint8_t CMD_SFRX = 0x3a;
static const uint8_t CMD_SFTX = 0x3b;

static const uint8_t REG_PARTNUM = 0xf0;
static const uint8_t REG_VERSION = 0xf1;

static const uint8_t CREG_FIFOTHR = 0x03;
static const uint8_t CREG_PKTLEN = 0x06;
static const uint8_t CREG_PKTCTRL1 = 0x07;
static const uint8_t CREG_PKTCTRL0 = 0x08;
static const uint8_t CREG_FREQ2 = 0x0d;
static const uint8_t CREG_FREQ1 = 0x0e;
static const uint8_t CREG_FREQ0 = 0x0f;
static const uint8_t CREG_MDMCFG4 = 0x10;
static const uint8_t CREG_MDMCFG3 = 0x11;
static const uint8_t CREG_MDMCFG2 = 0x12;
static const uint8_t CREG_MDMCFG1 = 0x13;
static const uint8_t CREG_MDMCFG0 = 0x14;
static const uint8_t CREG_MCSM0 = 0x18;
static const uint8_t CREG_FOCCFG = 0x19;
static const uint8_t CREG_FREND0 = 0x22;

static const uint8_t SREG_MARCSTATE = 0xf5;

static const uint8_t CREG_PATABLE_BURST_WRITE = 0x7e;
static const uint8_t CREG_PATABLE_BURST_READ = 0xfe;
static const uint8_t CFIFO_TX_BURST = 0x7f;

static const uint8_t MARCSTATE_TX = 0x13;
static const uint8_t MARCSTATE_TX_END = 0x14;
static const uint8_t MARCSTATE_RXTX_SWITCH = 0x15;

static const uint8_t MSG_SEND_ERRORS_RESET_LIMIT = 3;

/* Each symbol takes 220us. Computherm/Delta Q7RF uses PWM modulation.
   Every data bit is encoded as 3 bit inside the buffer.
   001 = 1, 011 = 0, 111000111 = preamble */
static const uint8_t Q7RF_REG_CONFIG[] = {
    CREG_FIFOTHR,  0x00,  // TX FIFO length = 61, others default
    CREG_PKTLEN,   0x3d,  // 61 byte packets
    CREG_PKTCTRL1, 0x00,  // Disable RSSI/LQ payload sending, no address check
    CREG_PKTCTRL0, 0x01,  // variable packet length, no CRC calculation
    CREG_FREQ2,    0x21,  // FREQ2, FREQ=0x216544 => 2.188.612 * (26 MHz OSC / 2^16) ~= 868.285 MHz
    CREG_FREQ1,    0x65,  // ^FREQ1
    CREG_FREQ0,    0x44,  // ^FREQ0
    CREG_MDMCFG4,  0xf7,  // baud exponent = 7 (lower 4 bits)
    CREG_MDMCFG3,  0x6B,  // baud mantissa = 107 -> (((256+107) * 2^7)/2^28) * 26 MHz OSC = 4.5 kBaud
    CREG_MDMCFG2,  0x30,  // DC filter on, ASK/OOK modulation, no manchester coding, no preamble/sync
    CREG_MDMCFG1,  0x00,  // no FEC, channel spacing exponent = 0 (last two bit)
    CREG_MDMCFG0,  0xf8,  // channel spacing mantissa = 248 -> 6.000.000 / 2^18 * (256 + 248) * 2^0 ~= 50kHz
    CREG_MCSM0,    0x10,  // autocalibrate synthesizer when switching from IDLE to RX/TX state
    CREG_FOCCFG,   0x00,  // ASK/OOK has no frequency offset compensation
    CREG_FREND0,   0x11   // ASK/OOK PATABLE (power level) settings = up to index 1
};

// 0xc0 = +12dB max power setting
static const uint8_t Q7RF_PA_TABLE[] = {0x00, 0xc0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};

unsigned long elapsed(unsigned long since, unsigned long now) {
  if (since > now) {
    // millis() overflows every ~50 days
    return (ULONG_MAX - since) + now;
  } else {
    return now - since;
  }
}

uint8_t state_to_msg(bool state) { return state ? MSG_HEAT_ON : MSG_HEAT_OFF; }

void encode_bits(uint16_t byte, uint8_t pad_to_length, char **dest) {
  char binary[9];
  itoa(byte, binary, 2);
  int binary_len = strlen(binary);

  if (binary_len < pad_to_length) {
    for (int p = 0; p < pad_to_length - binary_len; p++) {
      strncpy(*dest, Q7RF_ZERO_BIT_DATA, strlen(Q7RF_ZERO_BIT_DATA));
      *dest += strlen(Q7RF_ZERO_BIT_DATA);
    }
  }

  for (int b = 0; b < binary_len; b++) {
    strncpy(*dest, binary[b] == '1' ? Q7RF_ONE_BIT_DATA : Q7RF_ZERO_BIT_DATA, strlen(Q7RF_ONE_BIT_DATA));
    *dest += strlen(Q7RF_ZERO_BIT_DATA);
  }
}

void compile_msg(uint16_t device_id, uint8_t cmd, uint8_t *msg) {
  char binary_msg[360];
  char *cursor = binary_msg;

  // Preamble
  char *preamble_start = cursor;
  strncpy(cursor, Q7RF_PREAMBLE_DATA, strlen(Q7RF_PREAMBLE_DATA));
  cursor += strlen(Q7RF_PREAMBLE_DATA);

  char *payload_start = cursor;

  // Command
  encode_bits(device_id, 16, &cursor);
  encode_bits(8, 4, &cursor);
  encode_bits(cmd, 8, &cursor);

  // Repeat the command once more
  strncpy(cursor, payload_start, cursor - payload_start);
  cursor += cursor - payload_start;

  // Add a gap
  strncpy(cursor, Q7RF_GAP_DATA, strlen(Q7RF_GAP_DATA));
  cursor += strlen(Q7RF_GAP_DATA);

  // Repeat the whole burst
  strncpy(cursor, preamble_start, cursor - preamble_start);
  cursor += cursor - preamble_start;

  // Convert msg to bytes
  cursor = binary_msg;  // Reset cursor
  uint8_t *cursor_msg = msg;
  char binary_byte[9];
  binary_byte[8] = '\0';
  for (int b = 0; b < 45; b++) {
    strncpy(binary_byte, cursor, 8);
    cursor += 8;
    *cursor_msg = strtoul(binary_byte, 0, 2);
    cursor_msg++;
  }

#if ESPHOME_LOG_LEVEL >= ESPHOME_LOG_LEVEL_VERBOSE
  // Assemble debug print
  char debug[91];
  cursor = debug;
  cursor_msg = msg;
  for (int b = 0; b < 45; b++) {
    sprintf(cursor, "%x", *cursor_msg);
    cursor += 2;
    cursor_msg++;
  }
  ESP_LOGV(TAG, "Encoded msg: 0x%02x as 0x%s", cmd, debug);
#endif
}

bool Q7RFSwitch::reset_cc() {
  // Chip reset sequence. CS wiggle (CC1101 manual page 45)
  this->disable();
  delayMicroseconds(5);
  this->enable();
  delayMicroseconds(10);
  this->disable();
  delayMicroseconds(41);

  this->send_cc_cmd(CMD_SRES);
  ESP_LOGD(TAG, "Issued CC1101 reset sequence.");

  // Read part number and version
  uint8_t partnum;
  this->read_cc_register(REG_PARTNUM, &partnum);

  uint8_t version;
  this->read_cc_register(REG_VERSION, &version);

  ESP_LOGI(TAG, "CC1101 found with partnum: %02x and version: %02x", partnum, version);

  // Setup config registers
  uint8_t verify_value;
  for (int i = 0; i < sizeof(Q7RF_REG_CONFIG); i += 2) {
    this->write_cc_config_register(Q7RF_REG_CONFIG[i], Q7RF_REG_CONFIG[i + 1]);
    this->read_cc_config_register(Q7RF_REG_CONFIG[i], &verify_value);
    if (verify_value != Q7RF_REG_CONFIG[i + 1]) {
      ESP_LOGE(TAG, "Failed to write CC1101 config register. reg: %02x write: %02x read: %02x", Q7RF_REG_CONFIG[i],
               Q7RF_REG_CONFIG[i + 1], verify_value);
      return false;
    }
#if ESPHOME_LOG_LEVEL >= ESPHOME_LOG_LEVEL_VERBOSE
    ESP_LOGV(TAG, "Written CC1101 config register. reg: %02x value: %02x", Q7RF_REG_CONFIG[i], Q7RF_REG_CONFIG[i + 1]);
#endif
  }

  // Write PATable
  uint8_t pa_table[sizeof(Q7RF_PA_TABLE)];
  for (int i = 0; i < sizeof(Q7RF_PA_TABLE); i++)
    pa_table[i] = Q7RF_PA_TABLE[i];

  this->enable();
  this->write_byte(CREG_PATABLE_BURST_WRITE);
  this->write_array(pa_table, sizeof(Q7RF_PA_TABLE));
  this->disable();

  this->enable();
  this->write_byte(CREG_PATABLE_BURST_READ);
  this->read_array(pa_table, sizeof(Q7RF_PA_TABLE));
  this->disable();

  for (int i = 0; i < sizeof(Q7RF_PA_TABLE); i++) {
    if (pa_table[i] != Q7RF_PA_TABLE[i]) {
      ESP_LOGE(TAG, "Failed to write CC1101 PATABLE.");
      return false;
    }
#if ESPHOME_LOG_LEVEL >= ESPHOME_LOG_LEVEL_VERBOSE
    ESP_LOGV(TAG, "Written CC1101 PATABLE[%d]: %02x", i, Q7RF_PA_TABLE[i]);
#endif
  }

  return true;
}

void Q7RFSwitch::send_cc_cmd(uint8_t cmd) {
  this->enable();
  this->transfer_byte(cmd);
  this->disable();
}

void Q7RFSwitch::read_cc_register(uint8_t reg, uint8_t *value) {
  this->enable();
  this->transfer_byte(reg);
  *value = this->transfer_byte(0);
  this->disable();
}

void Q7RFSwitch::read_cc_config_register(uint8_t reg, uint8_t *value) { this->read_cc_register(reg + 0x80, value); }

void Q7RFSwitch::write_cc_register(uint8_t reg, uint8_t *value, size_t length) {
  this->enable();
  this->transfer_byte(reg);
  this->transfer_array(value, length);
  this->disable();
}

void Q7RFSwitch::write_cc_config_register(uint8_t reg, uint8_t value) {
  uint8_t arr[1] = {value};
  this->write_cc_register(reg, arr, 1);
}

bool Q7RFSwitch::send_cc_data(const uint8_t *data, size_t length) {
  uint8_t buffer[length];
  for (int i = 0; i < length; i++) {
    buffer[i] = *data;
    data++;
  }

  this->send_cc_cmd(CMD_SIDLE);
  this->send_cc_cmd(CMD_SFRX);
  this->send_cc_cmd(CMD_SFTX);

  this->enable();
  this->write_byte(CFIFO_TX_BURST);
  this->write_array(buffer, length);
  this->disable();

  this->send_cc_cmd(CMD_STX);

  uint8_t state;
  this->read_cc_register(SREG_MARCSTATE, &state);
  state &= 0x1f;
  if (state != MARCSTATE_TX && state != MARCSTATE_TX_END && state != MARCSTATE_RXTX_SWITCH) {
    ESP_LOGE(TAG, "CC1101 in invalid state after sending, returning to idle. State: 0x%02x", state);
    this->send_cc_cmd(CMD_SIDLE);
    return false;
  }

  return true;
}

bool Q7RFSwitch::send_msg(uint8_t msg) {
  if (msg == MSG_NONE)
    return false;

  bool result = false;
  const char *text_msg = NULL;
  switch (msg) {
    case MSG_HEAT_ON:
      result = this->send_cc_data(this->msg_heat_on_, sizeof(this->msg_heat_on_));
      text_msg = "HEAT ON";
      break;
    case MSG_HEAT_OFF:
      result = this->send_cc_data(this->msg_heat_off_, sizeof(this->msg_heat_off_));
      text_msg = "HEAT OFF";
      break;
    case MSG_PAIR:
      result = this->send_cc_data(this->msg_pair_, sizeof(this->msg_pair_));
      text_msg = "PAIR";
      break;
  }

  if (text_msg) {
    ESP_LOGD(TAG, "Sent message: %s", text_msg);
  }

  if (result) {
    this->last_msg_time_ = millis();
  } else {
    this->msg_errors_++;
    if (this->msg_errors_ >= MSG_SEND_ERRORS_RESET_LIMIT) {
      ESP_LOGE(TAG, "Multiple message send errors occured, forcing CC1101 reset.");
      this->reset_cc();
      this->msg_errors_ = 0;
    }
  }

  return result;
}

void Q7RFSwitch::set_state(bool state) {
  this->state_ = state;
  this->publish_state(state);
}

void Q7RFSwitch::setup() {
  // Revert switch to off state
  this->set_state(false);

  // Compile messages
  compile_msg(this->q7rf_device_id_, Q7RF_MSG_CMD_PAIR, this->msg_pair_);
  compile_msg(this->q7rf_device_id_, Q7RF_MSG_CMD_TURN_ON_HEATING, this->msg_heat_on_);
  compile_msg(this->q7rf_device_id_, Q7RF_MSG_CMD_TURN_OFF_HEATING, this->msg_heat_off_);

  // Register the pairing service
  register_service(&Q7RFSwitch::on_pairing, "q7rf_pair");

  this->spi_setup();
  if (this->reset_cc()) {
    ESP_LOGI(TAG, "CC1101 initialized.");
  } else {
    ESP_LOGE(TAG, "Failed to reset CC1101 modem. Check connection.");
    return;
  }

  this->initialized_ = true;
}

void Q7RFSwitch::on_pairing() {
  if (this->initialized_) {
    this->pending_msg_ = MSG_PAIR;
    ESP_LOGI(TAG, "Enqueued pairing.");
  }
}

void Q7RFSwitch::write_state(bool state) {
  if (this->initialized_) {
    if (state) {
      this->last_turn_on_time_ = millis();
    }

    if (this->state_ != state) {
      this->set_state(state);
      this->pending_msg_ = state_to_msg(state);
    }
  }
}

void Q7RFSwitch::dump_config() {
  ESP_LOGCONFIG(TAG, "Q7RF:");
  LOG_PIN("  CC1101 CS Pin: ", this->cs_);
  ESP_LOGCONFIG(TAG, "  Q7RF Device ID: 0x%04x", this->q7rf_device_id_);
  ESP_LOGCONFIG(TAG, "  Q7RF Resend interval: %d ms", this->q7rf_resend_interval_);
  ESP_LOGCONFIG(TAG, "  Q7RF Turn on watchdog interval: %d ms", this->q7rf_turn_on_watchdog_interval_);
}

void Q7RFSwitch::update() {
  if (this->initialized_) {
    if (this->pending_msg_ != MSG_NONE) {
      ESP_LOGD(TAG, "Handling prioritized message.");
      // Send prioritized message
      this->send_msg(this->pending_msg_);
      this->pending_msg_ = MSG_NONE;
    } else {
      unsigned long now = millis();

      if (this->state_ && this->q7rf_turn_on_watchdog_interval_ > 0 &&
          elapsed(this->last_turn_on_time_, now) > this->q7rf_turn_on_watchdog_interval_) {
        ESP_LOGD(TAG, "Turn on watch dog triggered, turning off furnace.");
        this->set_state(false);
        this->send_msg(MSG_HEAT_OFF);
        return;
      }

      // Check if we have to resend current state by now
      if (elapsed(this->last_msg_time_, now) > this->q7rf_resend_interval_) {
        ESP_LOGD(TAG, "Resending last state.");
        uint8_t msg = state_to_msg(this->state_);
        this->send_msg(msg);
      }
    }
  }
}

void Q7RFSwitch::set_q7rf_device_id(uint16_t id) { this->q7rf_device_id_ = id; }

void Q7RFSwitch::set_q7rf_resend_interval(uint32_t interval) { this->q7rf_resend_interval_ = interval; }

void Q7RFSwitch::set_q7rf_turn_on_watchdog_interval(uint32_t interval) {
  this->q7rf_turn_on_watchdog_interval_ = interval;
}

}  // namespace q7rf
}  // namespace esphome
